src/installation/guides/efi-stub.md

```diff
@@ -0,0 +1,162 @@
+packaged in Void, UEFI firmware can load the kernel directly as an EFI
+executable, eliminating the need for a bootloader.
+
+In the following, we set up a simple Void system on a drive, with an encrypted
+root partition and an unencrypted EFI system partition. We will refer to the
+drive as `/dev/sd<X>`.
+
+## Provisioning
+
+On the drive, create a GPT partition table with the following partitions:
+
+| No. | Size              | Type             |
+|-----|-------------------|------------------|
+| 1   | 512MiB            | EFI System       |
+| 2   | rest of the drive | Linux filesystem |
+
+Create a FAT 32 filesystem on the EFI System Partition, optionally giving it a
+label.
+
+```
+# mkfs.fat -F 32 -n efi /dev/sd<X>1
+```
+
+Initialize the LUKS device.
+
+```
+# cryptsetup luksFormat /dev/sd<X>2 --type luks2 --label luks
+```
+
+Open the LUKS device.
+
+```
+# cryptsetup open --type luks /dev/sd<X>2 root
+```
+
+Create the desired file system.
+
+```
+# mkfs.ext4 -L root /dev/mapper/root
+```
+
+## Installation
+
+With the LUKS device still open, mount the root filesystem, the EFI system
+partition and the necessary directories to `/mnt`.
+
+```
+# mount /dev/mapper/root /mnt
+# mkdir /mnt/boot
+# mount /dev/sd<X>1 /mnt/boot
+# for dir in dev proc sys run; do
+> mkdir /mnt/$dir
+> mount --rbind --make-rslave /$dir /mnt/$dir
+> done
+```
+
+Install the Void Linux base system, `cryptsetup` and `efibootmgr`.
+
+```
+# xbps-install -S -R https://alpha.de.repo.voidlinux.org/current -r /mnt base-system cryptsetup efibootmgr
+```
+
+Change the filesystem root to enter the new system.
+
+```
+# chroot /mnt
+```
+
+Set the filesystem permissions, root password and hostname.
+
+```
+# chown root:root /
+# chmod 755 /
+# passwd root
+# echo void > /etc/hostname
+```
+
+Void Linux uses [dracut](https://lwn.net/Articles/317793/) to manage early
+userspace. Configure it to open and mount the LUKS device and mount the EFI
+variables and system partition. To this end, edit `/etc/crypttab`.
+
+```
+# <name>    <device>                  <password>  <options>
+root        /dev/disk/by-uuid/<UUID>  none
+```
+
+The UUID of `/dev/sd<X>2` can be determined with `lsblk -f`. We'd like to write
+`/dev/disk/by-label/luks` instead, but it seems udev doesn't support LUKS labels
+yet. The password field is a path to a file containing the password or `none` to
+prompt during boot.
+
+Now, edit `/etc/fstab`.
+
+```
+# <file system>	        <dir> <type>  <options>             <dump>  <pass>
+tmpfs                   /tmp  tmpfs   defaults,nosuid,nodev 0       0
+efivarfs  /sys/firmware/efi/efivars  efivarfs  defaults     0       0
+/dev/disk/by-label/root /     ext4    defaults              0       1
+/dev/disk/by-label/efi  /boot vfat    defaults              0       2
+```
+
+Create `/etc/dracut.conf.d/30.conf` to configure dracut.
+
+```
+hostonly="yes"
+use_fstab="yes"
+install_items+=" /etc/crypttab "
+add_drivers+=" vfat nls_cp437 nls_iso8859_1 "
```

Comment:
What are the `nls` drivers?

Reply:
They are needed to mount the `vfat` filesystem. Otherwise, you'll see errors like 'codepage cp437 not found' or 'missing iocharset iso8859-1' in `dmesg`. It might be a locale problem, because I believe I could mount the filesystem correctly once void had booted, without those modules loaded. Is the mount in dracut somehow different from the one on the root filesystem?

Reply:
I don't know, but shouldn't be. The base system only has coreutils, no busybox or similar. Are those errors hard errors or warnings?

Reply:
Those are hard errors. They prevent the device from being mounted, that is, not the boot in general.

Reply:
Got it. They *are* loaded in my sytem too.

Reply:
Should I mention their purpose in the documentation?

Reply:
codepage 437 is the pc bios's native character encoding and iso 8859-1 is (roughly) the latin1 character encoding microsoft was using ca windows 95/98. They're both pretty much ascii, but the driver is probably using cp437 to decode the classic 8.3 filename and using iso 8859-1 to decode the extended filename.

As far as I can tell, the reason for this is the next step--- dracut is being configured to mount all the filesystems, instead of only mounting the root filesystem. This is explained further on to be to solve the problem of the post-dracut userland trying to re-open the luks device, but that's exactly why you're not supposed to put `/` in crypttab, right?

Reply:
I understand you'd prefer dracut to mount only the root?

Reply:
yeah, unless there's something I've missed. It has the disadvantage that modules for every filesystem mounted at boot need to go into the initramfs, the boot process needs to be modified, xbps needs to be configured to keep that change from being reverted. The only advantage is that you can use crypttab for configuration.

By the way, I'm also of the opinion that this should only cover efistub stuff, without worrying about disk encryption. we have a guide on full disk encryption, but I think a lot of it should be completely rewritten anyway, so these discussions are still important.

Reply:
> It has the disadvantage that modules for every filesystem mounted at boot need to go into the initramfs, the boot process needs to be modified, xbps needs to be configured to keep that change from being reverted. The only advantage is that you can use crypttab for configuration.

Okay, I agree.

> By the way, I'm also of the opinion that this should only cover efistub stuff, without worrying about disk encryption. We have a guide on full disk encryption, but I think a lot of it should be completely rewritten anyway, so these discussions are still important.

But where does efistub-specific documentation about disk encryption go? I think it belongs rather in the efistub documentation than any documentation related to encryption, because in my opinion one would want encryption most of the time and the documentation on encryption should not be an enumeration of sections whose relevance is mutually exclusive.